Add OIDC authentication Integration Tests

jkakavas opened the pull request:
This change introduces an OpenID Connect Provider in idp-fixture
that uses the existing openldap server for authentication and
as a user info repository.

It also adds tests where a mock facilitator application (HttpClient)
performs authentication using Elasticsearch's oidc realm against
that OpenID Connect Provider, using both the authorization code
and the implicit grant flows.

jkakavas committed:
Add OIDC authentication Integration Tests

This change introduces an OpenID Connect Provider in idp-fixture
that uses the existing openldap server for authentication and
as a user info repository.

It also adds tests where a mock facilitator application (HttpClient)
performs authentication using Elasticsearch's oidc realm against
that OpenID Connect Provider, using both the authorization code
and the implicit grant flows.

jkakavas committed:
Merge branch 'feature-oidc-realm' into oidc-realm-IT

elasticmachine commented:
Pinging @elastic/es-security

jkakavas committed:
remove superfluous config examples

jkakavas commented:
@elasticmachine test this please

jkakavas committed:
Replace java mail with jakarta mail (see #40088)

jkakavas committed:
update jar SHA

jaymode commented on x-pack/test/idp-fixture/docker-compose.yml:
```diff
@@ -38,3 +38,15 @@ services:
+    depends_on:
+      - openldap
+    ports:
+      - "8080:8080"
```
can we avoid the hardcoded port? See #40333 for my suggestion on how to do it for the rest of the fixture.

jkakavas replied:
yes! Will do :) I wasn't aware this could be handled this way, thanks!

jkakavas committed:
Merge branch 'feature-oidc-realm' into oidc-realm-IT

jkakavas commented on x-pack/qa/oidc-op-tests/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthIT.java:
```diff
@@ -0,0 +1,386 @@
+        adminClient().performRequest(createUserRequest);
+    }
+
+    @Before
```
I prefer to have different methods performing different logical tasks, irrespective to the fact that they all need to run before each test. I believe we use this pattern in many places and I don't see any particular problem with it ( especially granted there is no order of execution requirements here ). If you have strong opinions on the one @Before per class, I could add a third method , annotate that with @Before and simply call the two existing ones there, WDYT ?

jkakavas committed:
address feedback

jkakavas committed:
Merge branch 'feature-oidc-realm' into oidc-realm-IT

jkakavas committed:
Adjust after TokenService refactoring

jkakavas committed:
add necessary files for jakarta.mail

jkakavas committed:
use ephemeral ports instead of hardcoded ones for oidc-provider

jkakavas committed:
Adjust time fixing in tests to reduce possibility for random failures

atorok commented on x-pack/qa/oidc-op-tests/build.gradle:
```diff
@@ -0,0 +1,83 @@
+Project idpFixtureProject = xpackProject("test:idp-fixture")
+evaluationDependsOn(idpFixtureProject.path)
```
This is unnecessary.  `testFixtures.useFixture` does this behind the scenes

atorok commented on x-pack/qa/oidc-op-tests/build.gradle:
```diff
@@ -0,0 +1,83 @@
+task setupPorts {
+    dependsOn idpFixtureProject.postProcessFixture
+    doLast {
+        ephemeralPort = idpFixtureProject.postProcessFixture.ext."test.fixtures.oidc-provider.tcp.8080"
```
This won't work on hosts without docker `-Dtests.fixture.enabled=false` can be used to simulate it. 
I think it also needs a fix from  #40297

jkakavas replied:
I raised https://github.com/elastic/elasticsearch/pull/40585 because I wasn't sure how to do this properly and will be integrating whatever solution we come up with here to. Did you mean something like https://github.com/elastic/elasticsearch/pull/40297/files#diff-4fbfb78d054fdb66e0b02f321c8d63c5R79?

jkakavas replied:
> This won't work on hosts without docker 

These kind of projects only contain integration tests that depend on systems running on Docker so could we have a generic way of "Don't even consider this project if Docker is not available" instead of simply disabling the `integTest` task as I did in #40585 ?

alpar-t reviewed (commented):
I left some comments on the build side.

jkakavas committed:
address feedback

jkakavas committed:
Merge remote-tracking branch 'origin/feature-oidc-realm' into oidc-realm-IT

jaymode commented on x-pack/qa/oidc-op-tests/build.gradle:
```diff
@@ -0,0 +1,83 @@
+    setting 'xpack.security.authc.realms.oidc.c2id-implicit.claims.name', 'name'
+    setting 'xpack.security.authc.realms.oidc.c2id-implicit.claims.mail', 'email'
+    // Allow for troubleshooting CI errors
+    setting 'logger.org.elasticsearch.xpack.security.authc', 'TRACE'
```
Are you expecting this to fail in CI? I'd much rather leave this out because who knows what this could hide; maybe it slows things down enough that we miss a concurrency bug

jkakavas replied:
> Are you expecting this to fail in CI?

No :) , the idea was to have logs available for when it will - inevitably - fail . I see your point though, will remove this

jaymode commented on x-pack/qa/oidc-op-tests/src/test/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthIT.java:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    private CloseableHttpClient getHttpClient() {
+        return HttpClients.custom().build();
```
```suggestion
        return HttpClients.createDefault();
```

jaymode reviewed (changes requested):
I do not like the idea of checking in a web-app, `ldapAuth`. Is this something we could provision and/or build a image with it already there and just override the configuration?

jkakavas commented:
> I do not like the idea of checking in a web-app, ldapAuth. 

Well, we don't need to check the web app in - it is part of the image. The only reason we map this volume is so that we can replace the configuration of it (`ldapAuth/WEB-INF/ldapAuth.properties`) and since the authors of the image didn't make that configurable (like they did with `override.properties` for the c2id server ) I couldn't find a better way to do this. 
What I have tried : 

- try and map only the file with 
  ```
  - ./oidc/ldapAuth.properties:/c2id-server/tomcat/webapps/ldapauth/WEB-INF/ldapAuth.properties
  ```
  This doesn't work as when docker maps this, the `/c2id-server/tomcat/webapps/` is not yet created ( Tomcat hasn't still deployed the webapp ) and this means that we get `/c2id-server/tomcat/webapps/ldapauth` that only contains `/c2id-server/tomcat/webapps/ldapauth/WEB-INF/ldapAuth.properties` and no deployed application.

> Is this something we could provision and/or build a image with it already there and just override the configuration?

Yes I can build an image based on theirs and replace `ldapAuth.properties` in the Dockerfile. I tried to stay away from building the image at runtime (as @atorok mentioned this would be slower and thus not desirable in general ), but where would you see us hosting the docker image ?

jaymode commented:
Maybe there is a way to add a doLast on the postProcessFixture that calls `docker cp`?

jkakavas commented:
> Maybe there is a way to add a doLast on the postProcessFixture that calls docker cp?

That wouldn't work as ldapAuth would already be loaded by tomcat at that point and it doesn't reload config changes without restart. I'd need to see if the provided tomcat exposes its manager on 8080 and try and also do a curl to `localhost:ephemeralPort/manager/text/reload?path=/ldapAuth`

jkakavas committed:
Use c2id internal user for testing

This change makes it so we use the internal test user (alice) of
c2id image for testing. This in turn removes the need to configure
ldapAuth and the dependency to openldap.

jkakavas commented:
@jaymode I stopped using the ldapAuth application for our testing. What we care about in this context is the retrieved claims we get from the Provider, and not _how_ the provider sourced them (internal file, ldap, etc) . This means we don't have to deal with the issue of configuring ldapAuth anymore.

jaymode commented on x-pack/qa/oidc-op-tests/build.gradle:
```diff
@@ -0,0 +1,83 @@
+        ephemeralPort = idpFixtureProject.postProcessFixture.ext."test.fixtures.oidc-provider.tcp.8080"
+    }
+// Don't attempt to get ephemeral ports when Docker is not available
+setupPorts.onlyIf { idpFixtureProject.postProcessFixture.enabled }
```
shouldn't this line be outside of the task itself?

jkakavas replied:
:man_facepalming: copy-paste woes.

atorok replied:
I think it's in the right place, but the indentation is off and `setupPorts` should not be repeated as it's still in the task configuration. It's good to have a single block that configures the task fully.

jkakavas replied:
@atorok you mean something like 

```
task setupPorts {
    onlyIf { idpFixtureProject.postProcessFixture.enabled }
    dependsOn idpFixtureProject.postProcessFixture
    doLast {
        ephemeralPort = idpFixtureProject.postProcessFixture.ext."test.fixtures.oidc-provider.tcp.8080"
    }
}
```

jkakavas replied:
Ok, now that I re-read your comment it seemed much more clear to me than the fist time around, I'll change that

jaymode reviewed (approved):
I left one question, otherwise LGTM

jkakavas committed:
address feedback

alpar-t reviewed (approved):
Left a comment, other than that the build LGTM

jkakavas committed:
move onlyIf in the task
